File: util/anim/MaAnim.java
package common.util.anim;

import common.io.InStream;
import common.io.OutStream;
import common.system.files.FileData;
import common.system.files.VFile;
import common.util.BattleStatic;
import common.util.Data;

import java.io.PrintStream;
import java.util.Queue;

public class MaAnim extends Data implements BattleStatic {

	public static MaAnim newIns(FileData f) {
		if(f == null) {
			return new MaAnim();
		}

		try {
			return new MaAnim(f.readLine());
		} catch (Exception e) {
			e.printStackTrace();
			return new MaAnim();
		}
	}

	public static MaAnim newIns(String str) {
		return new MaAnim(VFile.readLine(str));
	}

	public int n;
	public Part[] parts;

	public int max = 1, len = 1;

	public MaAnim() {
		n = 0;
		parts = new Part[0];
	}

	public MaAnim(Queue<String> qs) {
		qs.poll();
		qs.poll();
		n = Integer.parseInt(qs.poll().trim());
		parts = new Part[n];
		for (int i = 0; i < n; i++)
			parts[i] = new Part(qs);
		validate();
	}

	private MaAnim(MaAnim ma) {
		n = ma.n;
		parts = new Part[n];
		for (int i = 0; i < n; i++)
			parts[i] = ma.parts[i].clone();
		validate();
	}

	@Override
	public MaAnim clone() {
		return new MaAnim(this);
	}

	public void revert() {
		for (Part p : parts)
			if (p.ints[1] == 11)
				for (int[] move : p.moves)
					move[1] *= -1;
	}

	public void validate() {
		for (int i = 0; i < parts.length; i++) {
			if(parts[i].ints[1] == 2) {
				for(int j = 0; j < parts[i].moves.length; j++) {
					if(parts[i].moves[j][1] < 0) {
						parts[i].moves[j][1] = 0;
					}
				}
			}
		}

		max = 1;
		for (int i = 0; i < n; i++)
			if (parts[i].getMax() > max)
				max = parts[i].getMax();
		len = 1;
		for (int i = 0; i < n; i++)
			len = Math.max(len, parts[i].getMax());
	}

	public void write(PrintStream ps) {
		ps.println("[maanim]");
		ps.println("1");
		ps.println(parts.length);
		for (Part p : parts)
			p.write(ps);
	}

	protected void restore(InStream is) {
		n = is.nextInt();
		parts = new Part[n];
		for (int i = 0; i < n; i++) {
			parts[i] = new Part();
			parts[i].restore(is);
		}
		validate();
	}

	protected void update(float f, EAnimD<?> eAnim, boolean rotate) {
		if (rotate)
			f %= max + 1;

		if (f == 0) {
			for (EPart e : eAnim.ent)
				e.setValue();
		}

		for (int i = 0; i < n; i++) {
			int loop = parts[i].ints[2];
			int smax = parts[i].max;
			int fir = parts[i].fir;
			int lmax = smax - fir;

			boolean prot = rotate || loop == -1;

			float frame;

			if (prot) {
				int mf = loop == -1 ? smax : max + 1;

				frame = mf == 0 ? 0 : (f + parts[i].off) % mf;
			} else {
				frame = f + parts[i].off;
			}

			if (loop > 0 && lmax != 0) {
				if (frame > fir + loop * lmax) {
					parts[i].ensureLast(eAnim.ent);
					continue;
				}
				if (frame <= fir)
					;
				else if (frame < fir + loop * lmax)
					frame = fir + (frame - fir) % lmax;
				else
					frame = smax;
			}

			parts[i].update(frame, eAnim.ent);
		}

		eAnim.sort();
	}

	protected void write(OutStream os) {
		os.writeInt(n);
		for (Part p : parts)
			p.write(os);
	}

}

File: util/anim/MaAnim.kt
package common.util.anim

import common.CommonStatic
import common.io.InStream
import common.io.OutStream
import common.io.assets.Admin
import common.io.assets.Admin.StaticPermitted
import common.io.assets.AssetLoader
import common.io.assets.AssetLoader.AssetHeader
import common.io.assets.AssetLoader.AssetHeader.AssetEntry
import common.io.json.JsonEncoder
import common.io.json.Test
import common.io.json.Test.JsonTest_0.JsonD
import common.io.json.Test.JsonTest_2
import common.pack.Source.AnimLoader
import common.pack.Source.ResourceLocation
import common.pack.Source.SourceAnimLoader
import common.pack.Source.SourceAnimSaver
import common.pack.Source.Workspace
import common.pack.Source.ZipSource
import common.system.files.FileData
import common.util.BattleStatic
import common.util.Data
import common.util.stage.EStage
import common.util.stage.StageMap
import common.util.stage.StageMap.StageMapInfo
import common.util.unit.UnitLevel
import io.BCPlayer
import page.JL
import page.anim.AnimBox
import page.support.ListJtfPolicy
import page.support.SortTable
import page.view.ViewBox
import page.view.ViewBox.Conf
import page.view.ViewBox.Controller
import page.view.ViewBox.VBExporter
import java.io.PrintStream
import java.util.*
import java.util.function.Function

class MaAnim : Data, BattleStatic {
    var n: Int
    var parts: Array<Part?>
    var max = 1
    var len = 1

    constructor() {
        n = 0
        parts = arrayOfNulls(0)
    }

    constructor(qs: Queue<String>) {
        qs.poll()
        qs.poll()
        n = qs.poll().trim { it <= ' ' }.toInt()
        parts = arrayOfNulls(n)
        for (i in 0 until n) parts[i] = Part(qs)
        validate()
    }

    private constructor(ma: MaAnim) {
        n = ma.n
        parts = arrayOfNulls(n)
        for (i in 0 until n) parts[i] = ma.parts[i]!!.clone()
        validate()
    }

    override fun clone(): MaAnim {
        return MaAnim(this)
    }

    fun revert() {
        for (p in parts) if (p!!.ints[1] == 11) for (move in p.moves) move[1] *= -1
    }

    fun validate() {
        max = 1
        for (i in 0 until n) if (parts[i]!!.getMax() > max) max = parts[i]!!.getMax()
        len = 1
        for (i in 0 until n) if (parts[i]!!.getMax() - parts[i]!!.off > len) len = parts[i]!!.getMax() - parts[i]!!.off
    }

    fun write(ps: PrintStream) {
        ps.println("[maanim]")
        ps.println("1")
        ps.println(parts.size)
        for (p in parts) p.write(ps)
    }

    fun restore(`is`: InStream) {
        n = `is`.nextInt()
        parts = arrayOfNulls(n)
        for (i in 0 until n) {
            parts[i] = Part()
            parts[i]!!.restore(`is`)
        }
        validate()
    }

    fun update(f: Int, eAnim: EAnimD<*>, rotate: Boolean) {
        var f = f
        if (rotate) f %= max
        if (f == 0 || rotate && f % max == 0) {
            for (e in eAnim.ent) e.setValue()
        }
        for (i in 0 until n) {
            val loop = parts[i]!!.ints[2]
            val smax = parts[i]!!.max
            val fir = parts[i]!!.fir
            val lmax = smax - fir
            val prot = rotate || loop == -1
            var frame = 0
            if (prot) {
                val mf = if (loop == -1) smax else max
                frame = if (mf == 0) 0 else (f + parts[i]!!.off) % mf
                if (loop > 0 && lmax != 0) {
                    if (frame > fir + loop * lmax) {
                        parts[i]!!.ensureLast(eAnim.ent)
                        continue
                    }
                    if (frame <= fir) ; else if (frame < fir + loop * lmax) frame = fir + (frame - fir) % lmax else frame = smax
                }
            } else {
                frame = f + parts[i]!!.off
                if (loop > 0 && lmax != 0) {
                    if (frame > fir + loop * lmax) {
                        parts[i]!!.ensureLast(eAnim.ent)
                        continue
                    }
                    if (frame <= fir) ; else if (frame < fir + loop * lmax) frame = fir + (frame - fir) % lmax else frame = smax
                }
            }
            parts[i]!!.update(frame, eAnim.ent)
        }
        eAnim.sort()
    }

    fun write(os: OutStream) {
        os.writeInt(n)
        for (p in parts) p.write(os)
    }

    companion object {
        fun newIns(f: FileData): MaAnim {
            return try {
                MaAnim(f.readLine())
            } catch (e: Exception) {
                e.printStackTrace()
                MaAnim()
            }
        }

        fun newIns(str: String?): MaAnim {
            return CommonStatic.def.readSave<MaAnim>(str, Function { f: Queue<String>? ->
                f?.let { MaAnim(it) } ?: MaAnim()
            })
        }
    }
}
